### PR TITLE
Do not validate initial dropdown menu value

### LINKED ISSUE #64: Required dropdown menu should not be invalid by default
Don’t run validate for the initial value.

Validation should only run after user changes.

### PR DESCRIPTION
**Pull request by web-padawan**
Fixes #64

---
This change is [Reviewable](https://reviewable.io/reviews/vaadin/vaadin-dropdown-menu/66)

### CODE REVIEW

**Comment by platosha** [comment]
Reviewed 1 of 3 files at r1.
Review status: 1 of 3 files reviewed at latest revision, 1 unresolved discussion.

---

*[vaadin-dropdown-menu.html, line 452 at r1](https://reviewable.io:443/reviews/vaadin/vaadin-dropdown-menu/66#-Kz8OCrEOo6YjBkwb8Gy:-Kz8OCrEOo6YjBkwb8Gz:b-44sb2l) ([raw file](https://github.com/vaadin/vaadin-dropdown-menu/blob/4eca4066c0fc5afbaedb49e01357ebf8fec30851/vaadin-dropdown-menu.html#L452)):*
> ```HTML
> 
>           // Skip validation for initial value
>           if (this._initialized === undefined) {
> ```

Introducing a flag for that is sad.

How about removing the `.validate()` call completely from here?

Perhaps it’s enough to call `.validate()` only on close (there is such a call already). And if not, perhaps we should identify these cases and have `.validate()` called explicitly for them, not as a side-effect of `_updateValueSlot()`.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/vaadin/vaadin-dropdown-menu/66)*

**Comment by web-padawan** [comment]
Review status: 0 of 3 files reviewed at latest revision, 1 unresolved discussion.

---

*[vaadin-dropdown-menu.html, line 452 at r1](https://reviewable.io:443/reviews/vaadin/vaadin-dropdown-menu/66#-Kz8OCrEOo6YjBkwb8Gy:-Kz9-NM63ACUkb16Sv3U:b-3mggf5) ([raw file](https://github.com/vaadin/vaadin-dropdown-menu/blob/4eca4066c0fc5afbaedb49e01357ebf8fec30851/vaadin-dropdown-menu.html#L452)):*
<details><summary><i>Previously, platosha (Anton Platonov) wrote…</i></summary><blockquote>

Introducing a flag for that is sad.

How about removing the `.validate()` call completely from here?

Perhaps it’s enough to call `.validate()` only on close (there is such a call already). And if not, perhaps we should identify these cases and have `.validate()` called explicitly for them, not as a side-effect of `_updateValueSlot()`.
</blockquote></details>

Removed the flag and refactored using single property observer. Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/vaadin/vaadin-dropdown-menu/66)*

**Comment by manolo** [comment]
It's missing to call validate when blurring the input. Take a look to the text-field required demo, it goes red when you focus and blur the input.

---

Review status: 0 of 4 files reviewed at latest revision, 2 unresolved discussions, some commit checks failed.

---

*[vaadin-dropdown-menu.html, line 421 at r2](https://reviewable.io:443/reviews/vaadin/vaadin-dropdown-menu/66#-Kz9dmZtayi1o5v0HE52:-Kz9dmZtayi1o5v0HE53:b-72vfd6) ([raw file](https://github.com/vaadin/vaadin-dropdown-menu/blob/6649ae79860a95b16f1220487e00918db512223a/vaadin-dropdown-menu.html#L421)):*
> ```HTML
>           this.validate();
>         }
> ```

I think it should be easier if you observe to this

```
'_valueChanged(value, opened)'
```

Then the code might be 
```
_valueChanged(value, opened) {
      opened && this.validate();
}
```

And the `this.validate()` could be removed from the opened observer.

I think you need to move `this.opened = false ` to the end of the `_updateValueSlot` block (after the line where value is changed`

Can you try?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/vaadin/vaadin-dropdown-menu/66#-:-Kz9dAC41_j8PttMOERb:bmzwwre)*

**Comment by web-padawan** [comment]
Review status: 0 of 3 files reviewed at latest revision, 2 unresolved discussions.

---

*[vaadin-dropdown-menu.html, line 421 at r2](https://reviewable.io:443/reviews/vaadin/vaadin-dropdown-menu/66#-Kz9dmZtayi1o5v0HE52:-KzNbrpVE6hMTRVRo46-:b-jcf9mn) ([raw file](https://github.com/vaadin/vaadin-dropdown-menu/blob/6649ae79860a95b16f1220487e00918db512223a/vaadin-dropdown-menu.html#L421)):*
<details><summary><i>Previously, manolo (Manuel Carrasco Moñino) wrote…</i></summary><blockquote>

I think it should be easier if you observe to this

```
'_valueChanged(value, opened)'
```

Then the code might be 
```
_valueChanged(value, opened) {
      opened && this.validate();
}
```

And the `this.validate()` could be removed from the opened observer.

I think you need to move `this.opened = false ` to the end of the `_updateValueSlot` block (after the line where value is changed`

Can you try?


</blockquote></details>

I tried that and it does not cover the case when `value` is set from outside to an empty string. 

So basically there are 3 cases when we need to validate:
1. value is changed (either by selecting item or from outside)
2. overlay is closed (especially when no value selected, otherwise it is covered by above case)
3. focus leaves the input

Added the implementation for 3rd case and tests.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/vaadin/vaadin-dropdown-menu/66)*

**Comment by platosha** [comment]
Reviewed 1 of 4 files at r2, 2 of 3 files at r3.
Review status: all files reviewed at latest revision, 2 unresolved discussions, some commit checks failed.

---

*[vaadin-dropdown-menu.html, line 299 at r1](https://reviewable.io:443/reviews/vaadin/vaadin-dropdown-menu/66#-KzOojaYNZzh2psREW8r:-KzOojaYNZzh2psREW8s:b-2a9u6u) ([raw file](https://github.com/vaadin/vaadin-dropdown-menu/blob/4eca4066c0fc5afbaedb49e01357ebf8fec30851/vaadin-dropdown-menu.html#L299)):*
> ```HTML
>             _inputElement: Object,
> 
>             _valueElement: Object,
> ```

Accidental removal? `_valueElement` is still in use.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/vaadin/vaadin-dropdown-menu/66)*

**Comment by web-padawan** [comment]
Review status: all files reviewed at latest revision, 2 unresolved discussions, some commit checks failed.

---

*[vaadin-dropdown-menu.html, line 299 at r1](https://reviewable.io:443/reviews/vaadin/vaadin-dropdown-menu/66#-KzOojaYNZzh2psREW8r:-KzOqKFXHrAji-80KPoB:b-2nqzfq) ([raw file](https://github.com/vaadin/vaadin-dropdown-menu/blob/4eca4066c0fc5afbaedb49e01357ebf8fec30851/vaadin-dropdown-menu.html#L299)):*
<details><summary><i>Previously, platosha (Anton Platonov) wrote…</i></summary><blockquote>

Accidental removal? `_valueElement` is still in use.
</blockquote></details>

This is not a removal, but vice versa: I added this property (it wasn't declared in `properties` block before).

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/vaadin/vaadin-dropdown-menu/66)*

**Comment by platosha** [comment]
Review status: all files reviewed at latest revision, 2 unresolved discussions, some commit checks failed.

---

*[vaadin-dropdown-menu.html, line 299 at r1](https://reviewable.io:443/reviews/vaadin/vaadin-dropdown-menu/66#-KzOojaYNZzh2psREW8r:-KzOtA1Fy4p24HBSXgfU:b-g2mhd4) ([raw file](https://github.com/vaadin/vaadin-dropdown-menu/blob/4eca4066c0fc5afbaedb49e01357ebf8fec30851/vaadin-dropdown-menu.html#L299)):*
<details><summary><i>Previously, web-padawan (Sergey Kulikov) wrote…</i></summary><blockquote>

This is not a removal, but vice versa: I added this property (it wasn't declared in `properties` block before).
</blockquote></details>

Oh, sorry, was viewing the wrong revision.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/vaadin/vaadin-dropdown-menu/66)*

**Comment by platosha** [comment]
:lgtm:

---

Review status: all files reviewed at latest revision, 1 unresolved discussion, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/vaadin/vaadin-dropdown-menu/66#-:-KzPGeOyYCT3Royp24aS:bnfp4nl)*

**Comment by web-padawan** [comment]
Rebased.

---

Reviewed 4 of 4 files at r4.
Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/vaadin/vaadin-dropdown-menu/66#-:-KzUVjLKpemB-zUlqCbk:bncvegj)*
